### PR TITLE
Drop python 3.6/3.7, remove unneeded compat

### PR DESCRIPTION
**Pull request by adrianeboyd**
## Description

Drop python 3.6/3.7, remove unneeded compat.

### Types of change

?

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by adrianeboyd** [comment]
There is one place where `spacy-legacy` uses the compat import, so this requires a v4.0.0.dev0 release of `spacy-legacy`: https://github.com/explosion/spacy-legacy/pull/38, https://github.com/explosion/spacy-legacy/pull/40

File: spacy/pipeline/spancat.py [hunk 1]
```diff
@@ -1,12 +1,11 @@
 from typing import List, Dict, Callable, Tuple, Optional, Iterable, Any, cast
-from typing import Union
+from typing import Union, Literal, Protocol, runtime_checkable
 from thinc.api import Config, Model, get_current_ops, set_dropout_rate, Ops
 from thinc.api import Optimizer
 from thinc.types import Ragged, Ints2d, Floats2d
 
 import numpy
 
-from ..compat import Protocol, runtime_checkable
```
Comment (polm) [inline spacy/pipeline/spancat.py:9]: Looks like you don't need Literal here?

**Review by polm — COMMENTED** [review]
Looks good to me except for one extra import.

Separately, this reminded me that we can use `shlex.join` now instead of `spacy.util.join_command`. That function itself will go away when the #11378 successor is merged, but might as well update it for now.

**Comment by adrianeboyd** [comment]
This PR isn't intended to include all the updates and refactoring that would be possible for 3.8+, just to drop 3.6/3.7 from the setup/CI and remove code that is only related to <3.8 compatibility.

**Review by svlandeg — APPROVED** [review]

**Pull request merged**